File: framec_tests/cpp/state_stack/state_stack.cpp
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

#include <vector>
#include <any>
#include <stack>
using namespace std;
#include "../FrameLang/FrameLang.h"

//=============== Compartment ==============//

class StateStackCompartment
{
public:
    int state;
    
    StateStackCompartment(int state)
    {
        this->state = state;
    }
    
    std::unordered_map<std::string, std::any> stateArgs;
    std::unordered_map<std::string, std::any> stateVars;
    std::unordered_map<std::string, std::any> enterArgs;
    std::unordered_map<std::string, std::any> exitArgs;
    FrameEvent *_forwardEvent = nullptr;
};

class StateStack
{
private:
    StateStackCompartment *_compartment_;
    StateStackCompartment *_nextCompartment_;
    
    
    
public:
    StateStack()
    {
        // Create and intialize start state compartment.
        _state_ = static_cast<int>(StateStackState::A);
        // Create state stack.
        
        this->_stateStack_ = new stack<StateStackCompartment>();
        
        
        _compartment_ = new StateStackCompartment(this->_state_);
        _nextCompartment_ = nullptr;
        
        
        // Send system start event
        FrameEvent frame_event(">", std::unordered_map<std::string, std::any>());
        _mux_(&frame_event);
    
    }
    
    // states enum
private:
    enum class StateStackState
    {
        A = 0,
        B = 1,
        C = 2
    };
    
    //====================== Multiplexer ====================//

private:
    void _mux_(FrameEvent *e)
    {
        if(this->_compartment_->state == static_cast<int>(StateStackState::A))
        {
            this->_sA_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(StateStackState::B))
        {
            this->_sB_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(StateStackState::C))
        {
            this->_sC_(e);
        }
        
        
        if(this->_nextCompartment_ != nullptr)
        {
            StateStackCompartment *nextCompartment = this->_nextCompartment_;
            this->_nextCompartment_ = nullptr;
            if (nextCompartment->_forwardEvent != nullptr && 
                nextCompartment->_forwardEvent->_message == ">")
            {
                this->_mux_(new FrameEvent( "<", this->_compartment_->exitArgs));
                this->_compartment_ = nextCompartment;
                this->_mux_(nextCompartment->_forwardEvent);
            }
            else
            {
                this->_doTransition_(nextCompartment);
                if (nextCompartment->_forwardEvent != nullptr)
                {
                    this->_mux_(nextCompartment->_forwardEvent);
                }
            }
            nextCompartment->_forwardEvent = nullptr;
        }
    }
    
        
        //===================== Interface Block ===================//
    public:
        void to_a() {
            FrameEvent e("to_a", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void to_b() {
            FrameEvent e("to_b", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void to_c() {
            FrameEvent e("to_c", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void push() {
            FrameEvent e("push", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void pop() {
            FrameEvent e("pop", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void pop_change() {
            FrameEvent e("pop_change", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
    
    //===================== Machine Block ===================//

private:
    
    void _sA_(FrameEvent *e)
    {
        if (e->_message == ">") {
            log_do("A:>");
            return;
        }
        else if (e->_message == "<") {
            log_do("A:<");
            return;
        }
        else if (e->_message == "to_a") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::A));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_b") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::B));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_c") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::C));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "push") {
            _stateStack_push_(_compartment_);
            return;
        }
        else if (e->_message == "pop") {
            StateStackCompartment *compartment = this->_stateStack_pop_();
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "pop_change") {
            StateStackCompartment *compartment = this->_stateStack_pop_();
            this->_changeState_(compartment);
            return;
        }
    }
    
    void _sB_(FrameEvent *e)
    {
        if (e->_message == ">") {
            log_do("B:>");
            return;
        }
        else if (e->_message == "<") {
            log_do("B:<");
            return;
        }
        else if (e->_message == "to_a") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::A));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_b") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::B));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_c") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::C));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "push") {
            _stateStack_push_(_compartment_);
            return;
        }
        else if (e->_message == "pop") {
            StateStackCompartment *compartment = this->_stateStack_pop_();
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "pop_change") {
            StateStackCompartment *compartment = this->_stateStack_pop_();
            this->_changeState_(compartment);
            return;
        }
    }
    
    void _sC_(FrameEvent *e)
    {
        if (e->_message == ">") {
            log_do("C:>");
            return;
        }
        else if (e->_message == "<") {
            log_do("C:<");
            return;
        }
        else if (e->_message == "to_a") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::A));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_b") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::B));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_c") {
            StateStackCompartment *compartment =  new StateStackCompartment(static_cast<int>(StateStackState::C));
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "push") {
            _stateStack_push_(_compartment_);
            return;
        }
        else if (e->_message == "pop") {
            StateStackCompartment *compartment = this->_stateStack_pop_();
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "pop_change") {
            StateStackCompartment *compartment = this->_stateStack_pop_();
            this->_changeState_(compartment);
            return;
        }
    }

//===================== Actions Block ===================//
    
    
    
public:
    void log_do(const std::string& msg)
    {
        tape.push_back(msg);
    }
    
    // Unimplemented Actions
    public:
    
    //===================== Domain Block ===================//
    
    
    public:
    std::vector<std::string> tape ;
    
    
    //=============== Machinery and Mechanisms ==============//

private:
    int _state_;
    
    void _transition_(StateStackCompartment *compartment)
    {
        _nextCompartment_ = compartment;
    }
    
    void _doTransition_(StateStackCompartment *nextCompartment)
    {
        this->_mux_(new FrameEvent("<", _compartment_->exitArgs));
        this->_compartment_ = nextCompartment;
        this->_mux_(new FrameEvent(">", this->_compartment_->enterArgs));
    }
    
    private:
    stack<StateStackCompartment>* _stateStack_ = nullptr;
    
    public:
    void _stateStack_push_(StateStackCompartment* compartment)
    {
        _stateStack_->push(*compartment);
    }
    
    StateStackCompartment* _stateStack_pop_()
    {
        StateStackCompartment* copyCompartment = &_stateStack_->top();
        _stateStack_->pop();
        return copyCompartment;
    }
    
    private: void _changeState_(StateStackCompartment* compartment)
    {
        this->_compartment_ = compartment;
    }
    
    public:
    string state_info(){
        return std::to_string(_compartment_->state);
        }
        
};

/********************

class StateStackController : public StateStack
{
public:
	StateStackController() : StateStack() {}
};

********************/

File: framec_tests/javascript/src/state_stack/state_stack.js
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

function FrameEvent(message, parameters) {

    var that = {};
    that._message = message;
    that._parameters = parameters;
    that._return = null;
    return that;
    
}

class StateStack {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    #stateStack
    
    constructor () {
        
        // Create state stack.
        
        this.#stateStack = [];
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sA_;
        this.#compartment = new StateStackCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        this.tape = [];
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    to_a() {
        let e = FrameEvent("to_a",null);
        this.#mux(e);
    }
    
    to_b() {
        let e = FrameEvent("to_b",null);
        this.#mux(e);
    }
    
    to_c() {
        let e = FrameEvent("to_c",null);
        this.#mux(e);
    }
    
    push() {
        let e = FrameEvent("push",null);
        this.#mux(e);
    }
    
    pop() {
        let e = FrameEvent("pop",null);
        this.#mux(e);
    }
    
    pop_change() {
        let e = FrameEvent("pop_change",null);
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sA_:
                this.#sA_(e);
                break;
            case this.#sB_:
                this.#sB_(e);
                break;
            case this.#sC_:
                this.#sC_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sA_(e) {
        switch (e._message) {
            case ">":
                {
                this.log_do("A:>");
                
                return;
                }
                
            case "<":
                {
                this.log_do("A:<");
                
                return;
                }
                
            case "to_a":
                {
                let compartment =  new StateStackCompartment(this.#sA_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "to_b":
                {
                let compartment =  new StateStackCompartment(this.#sB_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "to_c":
                {
                let compartment =  new StateStackCompartment(this.#sC_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "push":
                {
                this.#stateStack_push(this.#compartment);
                
                return;
                }
                
            case "pop":
                {
                let compartment = this.#stateStack_pop()
                this.#transition(compartment)
                
                return;
                }
                
            case "pop_change":
                {
                let compartment = this.#stateStack_pop()
                this.#changeState(compartment)
                
                return;
                }
                
        }
    }
    
    #sB_(e) {
        switch (e._message) {
            case ">":
                {
                this.log_do("B:>");
                
                return;
                }
                
            case "<":
                {
                this.log_do("B:<");
                
                return;
                }
                
            case "to_a":
                {
                let compartment =  new StateStackCompartment(this.#sA_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "to_b":
                {
                let compartment =  new StateStackCompartment(this.#sB_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "to_c":
                {
                let compartment =  new StateStackCompartment(this.#sC_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "push":
                {
                this.#stateStack_push(this.#compartment);
                
                return;
                }
                
            case "pop":
                {
                let compartment = this.#stateStack_pop()
                this.#transition(compartment)
                
                return;
                }
                
            case "pop_change":
                {
                let compartment = this.#stateStack_pop()
                this.#changeState(compartment)
                
                return;
                }
                
        }
    }
    
    #sC_(e) {
        switch (e._message) {
            case ">":
                {
                this.log_do("C:>");
                
                return;
                }
                
            case "<":
                {
                this.log_do("C:<");
                
                return;
                }
                
            case "to_a":
                {
                let compartment =  new StateStackCompartment(this.#sA_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "to_b":
                {
                let compartment =  new StateStackCompartment(this.#sB_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "to_c":
                {
                let compartment =  new StateStackCompartment(this.#sC_);
                
                
                this.#transition(compartment);
                
                return;
                }
                
            case "push":
                {
                this.#stateStack_push(this.#compartment);
                
                return;
                }
                
            case "pop":
                {
                let compartment = this.#stateStack_pop()
                this.#transition(compartment)
                
                return;
                }
                
            case "pop_change":
                {
                let compartment = this.#stateStack_pop()
                this.#changeState(compartment)
                
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    log_do(msg) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    #stateStack_push(compartment) {
        this.#stateStack.push(this.#deepClone(compartment));
    }
    
    #stateStack_pop(){
        return this.#stateStack.pop();
    }
    
    // deepcopy function for compartment
    #deepClone(target) {
        let copy = {}
        walk(target, copy);
        return copy;
        
        function walk(target, copy) {
            for (let key in target) {
                let obj = target[key];
                if (obj instanceof Function) {
                    let value = obj;
                    add(copy, key, value);
                } else if (obj instanceof Array) {
                    let value = [];
                    let last = add(copy, key, value);
                    walk(obj, last);
                } else if (obj instanceof Object) {
                    let value = {};
                    let last = add(copy, key, value);
                    walk(obj, last);
                } else {
                    let value = obj;
                    add(copy, key, value);
                }
                
            }
        }
        
        function add(copy, key, value) {
            copy[key] = value;
            return copy[key];
        }
    }
    
    #changeState(compartment) {
        this.#compartment = compartment;
    }
    
    state_info() {
        return this.#compartment.state.name;
    }
    
    
};

//=============== Compartment ==============//

class StateStackCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class StateStackController extends StateStack {

	constructor() {
	  super()
	}
	log_do(msg) {}
};

********************/

module.exports = StateStack
